Sources/StreamChat/Repositories/SyncOperations.swift

```diff
@@ -0,0 +1,140 @@
+
+            syncRepository?.syncMissingEvents(
+                using: lastPendingConnectionDate,
+                channelIds: context.localChannelIds,
+                bumpLastSync: false
+            ) { result in
+                switch result {
+                case let .success(channelIds):
+                    context.synchedChannelIds = Set(channelIds)
+                    // As per our sync logic, we should keep the last connection date.
+                    database?.write { session in
+                        session.currentUser?.lastPendingConnectionDate = context.lastConnectionDate
+                    } completion: { _ in done(.continue) }
+                case let .failure(error):
+                    context.synchedChannelIds = Set([])
+                    done(error.shouldRetry ? .retry : .continue)
+                }
+            }
+        }
+    }
+}
+
+class WatchChannelOperation: AsyncOperation {
+    init(controller: ChatChannelController, context: SyncContext) {
+        super.init(maxRetries: 2) { [weak controller] done in
+            guard let controller = controller, controller.isAvailableOnRemote else {
+                done(.continue)
+                return
+            }
+
+            // Reset only if it needs recovery
+            if let cid = controller.cid, context.synchedChannelIds.contains(cid) {
+                done(.continue)
+                return
+            }
+
+            let cidString = (controller.cid?.rawValue ?? "unknown")
+            log.info("Watching active channel \(cidString)", subsystems: .offlineSupport)
+            controller.watchActiveChannel { error in
+                if let cid = controller.cid, error == nil {
+                    log.info("Successfully watched active channel \(cidString)", subsystems: .offlineSupport)
+                    context.watchedChannelIds.insert(cid)
+                    done(.continue)
+                } else {
+                    let errorMessage = error?.localizedDescription ?? "missing cid"
+                    log.error("Failed watching active channel \(cidString): \(errorMessage)", subsystems: .offlineSupport)
+                    done(.retry)
+                }
+            }
+        }
+    }
+}
+
+class RefetchChannelListQueryOperation: AsyncOperation {
+    init(controller: ChatChannelListController, channelRepository: ChannelListUpdater, context: SyncContext) {
+        super.init(maxRetries: 2) { [weak controller] done in
+            guard let controller = controller, controller.isAvailableOnRemote else {
+                done(.continue)
+                return
+            }
+
+            log.info("Refetching channel lists queries & Cleaning up local message history", subsystems: .offlineSupport)
+            channelRepository.resetChannelsQuery(
+                for: controller.query,
+                watchedChannelIds: context.watchedChannelIds,
+                synchedChannelIds: context.synchedChannelIds
+            ) { result in
```

Comment:
Inside `resetChannelsQuery` we clean local history for channels that are not watched and are outside the 1st page. But as we discussed, we can't make a decision to clean up a channel until we sync all queries because a channel that matches Q1 and  Q2 can be outside the 1st page of Q1 but be in the 1st page of Q2. And if Q1 is synced first it will be cleaned up which is what we'd like to avoid.

Ideally, there should be a separate operation that runs as the last one in the chain.

Reply:
https://stream-io.atlassian.net/browse/CIS-1631